### ISSUE TITLE
long tests incorrectly killed

### ISSUE THREAD

**Issue by tdhock** [travis-ci/travis-ci#3849]
In the last few weeks Travis has started killing our tests since they take too long https://travis-ci.org/tdhock/animint/builds/57653159

```
* checking for unstated dependencies in ‘tests’ ... OK

* checking tests ...

Running ‘testthat.R’

No output has been received in the last 10 minutes, this potentially indicates a stalled build or something wrong with the build itself.

The build has been terminated
```

Is there a way to run R CMD check prefixed with `travis_wait` so that it won't be killed by travis during long tests? http://docs.travis-ci.com/user/build-timeouts/#Build-times-out-because-no-output-was-received

Or is there some other workaround?

**Comment by tdhock** [comment]
@craigcitro, @eddelbuettel and @hadley

**Comment by craigcitro** [comment]
hi @tdhock -- yeah, i should add this by default when running tests. patch coming in a minute.

**Comment by craigcitro** [comment]
actually, @BanzaiMan, quick question -- is the right thing to do here to just use `travis_wait` in the script we generate, or should i instead add `wait` support to `travis_cmd` (as another [option here](https://github.com/travis-ci/travis-build/blob/master/lib/travis/build/templates/header.sh#L19-L28))?

**Comment by BanzaiMan** [comment]
What is taking long? Is it expected to be silent for 10+ minutes? Is it possible to be more verbose by default?

`travis_wait` is a hack. It's better if you can avoid using that.

**Comment by tdhock** [comment]
I'm doing a bunch of unit tests for my R package, using a headless browser (RSelenium + phantomjs).

By default the R CMD check testing infrastructure only outputs 1 line per tests/*.R file, and all my tests occur in one such file.

I do not know if it is possible to have R CMD check output more verbose tests.

**Comment by BanzaiMan** [comment]
So, this long wait is dependent on the tests users run, correct? If so, I would think advising users to override the default commands is more appropriate than the default build script intervening.

**Comment by eddelbuettel** [comment]
None of the R unit testing frameworks force you to use just one file.  Splitting over several is trivial.  Can you just try doing that?

**Comment by tdhock** [comment]
Splitting over several tests/*.R files is not trivial when using testthat, which recommends one tests/testthat.R file which runs all the tests in tests/testthat/test-*.R files.

**Comment by eddelbuettel** [comment]
Then I misunderstood.   I am a user of RUnit, but doesn't testthat offer fancypants progress bars?  That would be enough to tickle output, no?

**Comment by cpsievert** [comment]
I think part of the issue might be due to recent changes in the reporter for `test_check()`. See the fifth bullet point [here](https://github.com/hadley/testthat/blob/48c231b1408a72963f65bc0b41c680c988839c78/NEWS.md#testthat-0919000)

**Comment by tdhock** [comment]
The problem is with R CMD check, which only outputs 1 line per `tests/*.R` file.

I asked R-devel and I got the following response from Peter Dalgaard. Is this possible?

```
If [travis] doesn't support setting parameters for the timeout, and it is anything Unix-like, can't you just set up a background process to echo a line every five minutes? As in

(for i in 1 2 3 4 ; do sleep 10 ; echo "ping" ; done ) & R ; kill %

(for larger values of 10, obviously; just showing the basic idea.)
```

**Comment by tdhock** [comment]
Another option would be for r-travis to NOT run R CMD check. For animint package purposes I think it should be fine to just run something like `R --no-save < tests/testthat.R` which would only run the tests (and show more output than R CMD check).

**Comment by tdhock** [comment]
Or is there an option to set the timeout to 100 minutes instead of 10 minutes?

**Comment by eddelbuettel** [comment]
The file `tests/testthat.R` is code.  Your code.  You can define whatever policy you desire, and eg set env vars which the different test scripts pick up and act upon.

Rcpp does something like that to reduce the number of tests running if it is a CRAN upload.  Those always have versions of the a.b.c form, where development uses a.b.c.d --- and in that case we run all tests.

**Comment by craigcitro** [comment]
@tdhock so to get you unblocked quickly, it looks like you have a few options:

1. slice up the tests into multiple files
2. use a different test reporter in `testthat`
3. specify your own `script` section in your `.travis.yml`
4. switch to using `r-travis`, which would give you a few extra knobs and levers.

i'd say one of the first two should be your quickest path forward?

**Comment by hadley** [comment]
@craigcitro I'm not sure a different test reporter will help - I think `R CMD check` caches all the output and only displays if there's an error.

**Comment by tdhock** [comment]
Thanks for the helpful suggestions @craigcitro 

option 1 (slice up the tests into multiple files) does not make sense with the testthat package.

option 2 (use a different test reporter in testthat) will not help since indeed R CMD check will not display any output other than if there is an error.

isn't option 4 (switch to using r-travis) deprecated? on the Travis R page http://docs.travis-ci.com/user/languages/r/ is says "you're encouraged to switch to using the native support"

I guess that just leaves us option 3. Is that OK with you @cpsievert ?

**Comment by craigcitro** [comment]
as to (4), you're definitely encouraged to move -- if it doesn't break you. `r-travis` won't go anywhere until i'm sure everyone is up and running on native support.

**Comment by cpsievert** [comment]
Thanks for the input everyone. Running tests outside of `R CMD check` will work for our purposes.

**Comment by cdeterman** [comment]
Where there any thoughts when the R CMD build doesn't complete within 10 minutes as opposed to during tests?  I have a package that has lots of C++ code that takes some time to compile so the build times out.

**Comment by hadley** [comment]
Why doesn't it produce output during building? C++ builds normally do

**Comment by cdeterman** [comment]
@hadley, I thought so too but for some reason it stops producing output during R CMD build when it is *checking if package can be installed*.  You can see my most recent build [here](https://travis-ci.org/cdeterman/gpuR).  I verified that this behavior is the same if I build locally with `devtools::build`.  It produces no output during this part although it eventually finishes on my local system.

**Comment by craigcitro** [comment]
@cdeterman is it actually the `./configure` step that it's hanging on? if so, could you force it to skip a few checks?

**Comment by cdeterman** [comment]
@craigcitro  My `configure` file is actually very simple right now so that actually surprises me.  It just sets some flags currently with the intent to enrich it down the line to have checks.  So there aren't any steps I can currently skip.

**Comment by craigcitro** [comment]
ah, then it's probably just the compilation steps. i don't have an easy answer there, unfortunately.

**Comment by jakirkham** [comment]
So, this sort of problem is not unique to R. For instance, `tox` has even more quiet behavior with no option that I know of to override it. The entire test suite will run without any output. While it could be argued that these testing suites should do things different, they work fine for users in most cases. Not to mention, it is asking for a significant amount of work on behalf of users to fix this problem. When a simple override on Travis CI's limits would more reliably solve this problem.

**Comment by levithatcher** [comment]
I love travis-ci, but am also frustrated this "No output has been received in the last 10m" error. It's happening while checking examples (which aren't numerous). Any update on this? Would be much appreciated!

**Comment by bbolker** [comment]
I'm stuck on a variant of this problem with an R package: https://travis-ci.org/lme4/lme4/builds/155465553 . [This comment above](https://github.com/travis-ci/travis-ci/issues/3849#issuecomment-98834630) more or less summarizes my situation, although r-travis is now deprecated, so the only option is to write my own `script:` section, but I'm stuck on how to do that.

Happy to re-post this as a separate issue if that seems better.

**tl;dr** is there a way to introduce a `travis_wait` statement in my `.travis.yml` for the `R CMD check stage`? I'm using the `language:r` approach and I'm a dummy, so I can't see an obvious way to do this. Pointers to any other solutions or workarounds would be *greatly* appreciated.

#5469 (closed) covers the same ground. That package `RClone` seems to have solved the problem (somehow), but I've been trawling through [their .travis.yml commit history](https://github.com/dbailleul/RClone/commits/master/.travis.yml) without success; they played around with non-native builds and seem to have ended up by going back to native builds, solving the timeout problem by disabling manual and vignette building (which won't solve my problem). I'm trying to cargo-cult my way through this, but seem to be failing pretty miserably. (My last test "passed" [here](https://travis-ci.org/lme4/lme4/builds/155470270), but only because I'm still getting the travis syntax wrong ...)

- we do have a lot of tests in the `testthat` framework, which only get reported as a single `Running ‘AAAtest-all.R'` line - this is where we get no output for 10 minutes and the build fails.
- until 2 weeks ago, Travis builds were still fine: https://travis-ci.org/lme4/lme4/builds . More recent commits haven't touched anything that should slow down the tests in any significant way (the commit where this problem first surfaced was a [one-line change to the README file](https://github.com/lme4/lme4/commit/2c54df2ce81231630be7a11be711ff7ac4827f93) ...) (Always possible, but it would be very surprising.) Our builds used to take 13-25 minutes total (usually <17 minutes). I don't know what's changed - something about the Travis build environment or upstream packages? (R CMD check on my laptop takes ~15 minutes total.)

**Comment by jgabry** [comment]
We're having a similar issue to @bbolker's with our [rstanarm](github.com/standev-rstanarm) package

**Comment by jimhester** [comment]
Try using the following, which uses [travis_wait()](https://docs.travis-ci.com/user/common-build-problems/#Build-times-out-because-no-output-was-received) function to print output every so often to avoid the timeout.

```yaml
script: -|
  R CMD build .
  travis_wait 20 R CMD check packageName*tar.gz
```

**Comment by MarkEdmondson1234** [comment]
I added:

```
script: 
  - |
    R CMD build .
    travis_wait 20 R CMD check googleComputeEngineR*tar.gz
```

..to the build and it [errored](https://travis-ci.org/cloudyr/googleComputeEngineR/builds/169261095) but losing the error log which I guess comes from doing the check the normal way.  I'm going to try splitting up the tests into different files, the problem is the tests call remote servers and wait for them to finish which is tipping it over the timeout.

**Comment by jimhester** [comment]
@MarkEdmondson1234 You can get the error logs with the following

```yaml
after_failure:
  find *Rcheck -name '*.fail' -print -exec cat '{}' \;
```

**Comment by jgabry** [comment]
@jimhester Thanks! Using `travis_wait 40` it appears to be successfully finishing R CMD check.

**Comment by MarkEdmondson1234** [comment]
Dear @jimhester, thanks for the command, it works in showing the errors and the tests now complete in the time, but it now timeouts when uploading the covr results - they are uploaded successfully but on Travis it doesn't return a success, so I'm in strange position of a successful covr but unsuccessful travis :)

https://travis-ci.org/cloudyr/googleComputeEngineR/builds/174732893

Is there anything I can do to force return?

**Comment by MarkEdmondson1234** [comment]
Dear @jimhester, thanks for the command, it works in showing the errors and the tests now complete in the time, but it now timeouts when uploading the covr results - they are uploaded successfully but on Travis it doesn't return a success, so I'm in strange position of a successful covr but unsuccessful travis :)

https://travis-ci.org/cloudyr/googleComputeEngineR/builds/174732893

Is there anything I can do to force return?

**Comment by mcdjork** [comment]
What is the recommended workaround for doing this using native r on travis-ci?

**Comment by pat-s** [comment]
After adding 

```
script: -|
  R CMD build .
  travis_wait 40 R CMD check packageName*tar.gz
```

I get the following build error

```
$ -| R CMD build . travis_wait 40 R CMD check sperrorest*.tar.gz
/home/travis/build.sh: line 57: eval: -|: invalid option
eval: usage: eval [arg ...]
The command "-| R CMD build . travis_wait 40 R CMD check sperrorest*.tar.gz" exited with 2.
```

[Build log](https://travis-ci.org/pat-s/sperrorest/jobs/195961309). What´s wrong in my setup?

**Comment by BanzaiMan** [comment]
@pat-s That's not what you added. Please review [your `script`](https://github.com/pat-s/sperrorest/blob/ae06d99df10a920e246eaa4e3afa6f6745dc3267/.travis.yml#L25-L28).

**Comment by pat-s** [comment]
@BanzaiMan See the modified attempt [here](https://github.com/pat-s/sperrorest/commit/a6ee714380f28a09fc073d50183580f313b9a8ea).

[Build](https://travis-ci.org/pat-s/sperrorest/jobs/196541990) still errors. Could you explain in more detail whats wrong here? 

```
$ -| R CMD build . travis_wait 40 R CMD check sperrorest*tar.gz
/home/travis/build.sh: line 57: eval: -|: invalid option
eval: usage: eval [arg ...]

```

**Comment by BanzaiMan** [comment]
@pat-s You need a space between `-` and `|` to introduce a literal block in YAML.  http://www.yaml.org/spec/1.2/spec.html

**Comment by pat-s** [comment]
@BanzaiMan thanks, did not know that.

@jimhester maybe you could update your [comment](https://github.com/travis-ci/travis-ci/issues/3849#issuecomment-255096774) and insert a whitespace? There maybe other users who might face the same problem otherwise

**Comment by pat-s** [comment]
@jimhester Is `R CMD build` here not accepting options? See [build log](https://travis-ci.org/pat-s/sperrorest/builds/196607591)

```
script: 
  - |
    R CMD build --no-build-vignettes.
    travis_wait 40 R CMD check --no--vignettes sperrorest*tar.gz
```

**Comment by jimhester** [comment]
@pat-s My comment does have a whitespace.

Your script is failing because you did not put a space between the `--no-build-vignettes` argument and `.`.

**Comment by BanzaiMan** [comment]
@jimhester I fixed your comment. ;-)

**Comment by rohan-shah** [comment]
I am having a similar problem to the ones listed here. My package is heavy on C++ and templates, and it seems that it may take more than 10 minutes to build. See https://travis-ci.org/rohan-shah/mpMap2/builds/199479721. 

Could this please be reopened? There is a need to incorporate travis_wait into the test and compilation stage.

**Comment by chrisknoll** [comment]
It would be nice if the native mode supported a travis_wait for CMD check and build instead of resorting to writing a custom script: section.

**Comment by achubaty** [comment]
I just bumped into this issue and came across a [nice solution](http://randyzwitch.com/travisci-10minute-timeout-build-error/) that allows one to split up `testthat` tests across multiple files using the `filter` argument in `test_check`. Basically, you replace the `tests/test-all.R` file, which looked like this:

```r
library(testthat)
test_check("myPkg")
```
with multiple test files (`tests/test-group1.R`, `tests/test-group2.R`, etc.), with each of these looking like this:

```r
library(testthat)
test_check("myPkg", filter = "fn1") 
# can include additional test_check lines to check fn2, fn3, etc. that are part of this test group
```

It's a bit more overhead to have those additional files in `tests/` acting as pointers to the files in `/tests/testthat/`, but you don't need to mess with working around how the tests are run and capturing/printing the log files.

**Comment by andrie** [comment]
It's a nice solution.  Just remember to update your `test-all.R` file if you add new tests!

**Comment by jimhester** [comment]
One way to set this up without needing to update for new test files is to use a regex for ranges of the alphabet.

e.g.

in `tests/a-m.R`
```r
library(testthat)
test_check("myPkg", filter = "^[a-m]")
```

in `tests/n-z.R`
```r
library(testthat)
test_check("myPkg", filter = "^[n-z]")
```

Then test files like (`test-apples.R`, `test-dogs.R`, `test-moons.R`) will be tested in the first file and (`test-queens.R`, `test-whales.R`, `test-zebras.R`) will be in the second.

**Comment by dfsp-spirit** [comment]
jimhester wrote:
> One way to set this up without needing to update for new test files is to use a regex for ranges of the alphabet.
> [...]
> Then test files like (`test-apples.R`, `test-dogs.R`, `test-moons.R`) will be tested in the first file and (`test-queens.R`, `test-whales.R`, `test-zebras.R`) will be in the second.

I did this and the maximum runtime over all these sets of tests is 3:46 min on my machine, but I still get the same timeout error on Travis in about every 3rd test run (most likely because their nodes are slower or have more load than my machine). And as far as I can tell, there is no way to tell which of the test sets (I have 8) is timing out.

Altogether this alone does not look like a great solution unless one was very close to the 10 minutes mark anyways, or is willing to split the tests into dozens of extra files, prefixed with different letters. I would suggest to combine this solution with the `travis_wait` command mentioned earlier in this thread.
